### ISSUE TITLE
sled-agent needs to deal with SSD overprovisioning

### ISSUE THREAD

**Issue by rmustacc** [oxidecomputer/omicron#4619]
With upcoming hardware BOM changes, sled-agent is going to need to take care of dealing with SSD overprovisioning and namespace reformatting changes prior to setting up and using the SSDs for any particular purpose. In particular, for the WDC SN655 devices, we will need to issue the vendor-specific resize command to the devices. This process erases all data on the device, so we'll need to think through the lifecycle carefully.

While this is focused on the U.2s, this is a more general problem in sled-agent where there are certain things we need to be doing on a per-model basis for all storage. For example, for the M.2s, right now we are relying on the factory to partition them and to reformat the namespace. For the SN840s, we have ended up with 512-byte namespaces instead of 4 KiB (though unclear how much of a difference that strictly makes or not).

Note, this all must occur before or as part of making these items into `ParsedDisks` or the M.2 equivalent. However, we have to be careful. As the original SN840s were not formatted to have 4 KiB namespaces, if they have data on them, we **must not** perform this transition. That may mean it only makes sense to look at the SN855 or field replacements for the Micron 7300/7450 M.2s.

To support this technologically, we're working through [illumos IPD 43 NVMe 2.0, libnvme, and the nvme(4D) ioctl interface](https://github.com/illumos/ipd/blob/master/ipd/0043/README.adoc), which will provide a programmatic libnvme interface, which while still private to illumos, sled agent will be able to consume to perform these operations.

**Comment by eanders01** [comment]
From the Mfg Test Meeting 22 Feb, I took an action to update on timing for this to land:

TLDR -- 6 Mar. More context below!

We are currently building 3 racks. REV 007 Gimlets have [new U.2 SSDs](https://mfg.duro.app/component/view/65835cee321c450007ad75f7) which will be incorporated into `P005`. This rack will be ready for racktest by 6 Mar.

Rack and Gimlet Configs:
```
PVT3       P004       P005
REV 005    REV 005    REV 007*
           REV 006
```
